### DISCUSSION TITLE
Non-integer charge amount

### DISCUSSION THREAD

**Discussion by povlhp** [goecharger/go-eCharger-API-v2#138] [Ideas]
One idea, is it possible to have a change that allows non-integer AMA value ?

Since it is a PWM signal - there is nothing that prevents the code from sending 6.5A on the line to the car. Not sure how the car would interpret it. 

I have done some digital to PWM before - but on a slow atmega/Arduino. so I know timing can be a pain. But I think the esp has native PWM

**Comment by 0xFEEDC0DE64** [comment]
Our V3 chargers cannot do non-integer current limits, V4 chargers could, but the firmware and charging logic are common, I will see what I can do.

**Comment by peterpoetzi** [comment]
Actually the hardware of V3 and V4 charger can do non-integer current settings, but so far we do not know if there are any cars that actually use a non-integer limit for charging

**Reply by povlhp** [reply]
If trivial to implement we could try. No need to expose in GUI. 

Not sure how the car will interpret it either. But would be interesting to try.

**Comment by povlhp** [comment]
One more thing, we need to preserve flash memory / nvram, so you would not have to store non-integer AMA value in flash, it makes no sense to do so, as we use continous values only for short time adjustment. Either ignore writing them completely, og write only the integer part (and only if changed). Or make 2 different values where the float is never stored.

**Reply by povlhp** [reply]
Got PV surplus charging running. Car is Kia EV6. 

Seems like response is pretty quick. So having a flag to enable it on PV mode would be great. 

![image](https://user-images.githubusercontent.com/11041820/233185949-63039a0f-0fbf-4541-8465-488b2ce63303.jpeg)

**Reply by 0xFEEDC0DE64** [reply]
if you manage to wear out our v3 or v4 flash you would unlock an archievment.
Changing the api keys no longer causes a flash erase anymore. Config keys are divided into pages on flash and only setting bits where needed.

**Comment by ehorvat1** [comment]
Hi, I also would love to have non integer a charging current .... are there any news on this? 
This would make PV surplus charging - specially in 3-phase mode - much better. 
And of corse this control value must not be saved to flash to save the flash from dying too quick.

**Reply by povlhp** [reply]
We don't know if cars will understand it, or round to nearest whole Ampere. It is an analog signal leaving room for some error.

**Reply by ehorvat1** [reply]
I am quite shure a car charger is not restricted to "whole" ampere charging setpoints. 
PWM Signals can be regardes to be more analog than digital so in no case the PWM amp. setpoint is spot on a integer ampere setpoint.
And there are some wallboxes which do not use integer ampere setpoins like: 
* Smartfox Pro charger (from Schrack ?) or
* KeContact P30 PV Edition (from Keba) see page 2 of: https://media.excentos.com/keba/pdf/de/kecontact-p30-pv-edition-produktflyer_2022_de_web.pdf

A good (german) explanation to the PWM signal on the CP contact of the Type 2 connector can be found here: https://evsim.gonium.net/
And there are "analog" Wallboxes which would use no Mocroprocessor at all, like described here: http://www.analogevse.xyz/AnalogEVSE-de.html

I dont know why Go-e guys think cars would only use integer values for charging ampere setpoints - makes absolutely no sense for me.
I think that is somehow historically grown....and easy to code into a PWM driver - which uses anyway a percent value for its duty cycle setpoint.

So instead a integer value for ampere setpoint a percent value of max. carging current would already help a lot (thats what Keba is doing....)
